vsphere/resource_vsphere_compute_cluster.go

```diff
@@ -1190,9 +1231,195 @@ func expandClusterConfigSpecEx(d *schema.ResourceData, version viapi.VSphereVers
+	return &hssProps, err
+}
+
+func generateDiskMap(client *govmomi.Client, host *object.HostSystem, list []interface{}) (*types.VsanHostDiskMapping, error) {
+	hssProps, err := hostStorageSystemPropertiesFromHostSystemID(client, host.Reference().Value)
+	diskMap := types.VsanHostDiskMapping{
+		NonSsd: []types.HostScsiDisk{},
+	}
+	if err != nil {
+		return nil, err
+	}
+	if hssProps.StorageDeviceInfo == nil {
+		return &diskMap, nil
+	}
+	if hssProps.StorageDeviceInfo == nil {
+		return &diskMap, nil
+	}
+	for _, scsiLun := range hssProps.StorageDeviceInfo.ScsiLun {
+		for _, diskGroup := range list {
+			if hostDisk, ok := scsiLun.(*types.HostScsiDisk); ok {
+				if err != nil {
+					return nil, err
+				}
+				for _, storageDisk := range diskGroup.(map[string]interface{})["storage"].(*schema.Set).List() {
+					if hostDisk.CanonicalName == storageDisk.(string) {
+						diskMap.NonSsd = append(diskMap.NonSsd, *hostDisk)
+					}
+				}
+				if diskGroup.(map[string]interface{})["cache"].(string) == hostDisk.CanonicalName {
+					diskMap.Ssd = *hostDisk
+				}
+			}
+		}
+	}
+	return &diskMap, nil
+}
+
+func deleteVsanDisks(host *object.HostSystem, list []interface{}, client *govmomi.Client) error {
+	log.Printf("deleteVsanDisks: Starting removal of vSAN disks on %s.", host.Name())
+	hvs, err := vsansystem.FromHost(client, host, defaultAPITimeout)
+	if err != nil {
+		return nil
+	}
+	diskMap, err := generateDiskMap(client, host, list)
+	if err != nil {
+		return err
+	}
+	if diskMap.Ssd.CanonicalName != "" || len(diskMap.NonSsd) > 0 {
+		log.Printf("deleteVsanDisks: Scheduled disks are being removed.")
+		if err = vsansystem.RemoveDiskMapping(client, host, hvs, diskMap, defaultAPITimeout); err != nil {
+			return err
+		}
+		log.Printf("deleteVsanDisks: vSAN disks successfully removed.")
+	} else {
+		log.Printf("deleteVsanDisks: No vSAN disks to remove on %s.", host.Name())
+	}
+	return nil
+}
+
+func addVsanDisks(host *object.HostSystem, list []interface{}, client *govmomi.Client) error {
+	log.Printf("addVsanDisks: Starting initialization of vSAN disks on %s.", host.Name())
+	hvs, err := vsansystem.FromHost(client, host, defaultAPITimeout)
+	if err != nil {
+		return nil
+	}
+	diskMap, err := generateDiskMap(client, host, list)
+	if diskMap.Ssd.CanonicalName != "" {
+		log.Printf("addVsanDisks: Scheduled disks are being initialized.")
+		if err = vsansystem.InitializeDisks(client, host, hvs, diskMap, defaultAPITimeout); err != nil {
+			return err
+		}
+		log.Printf("addVsanDisks: vSAN disks successfully initialized.")
+	} else {
+		log.Printf("addVsanDisks: No vSAN disks to initialize on %s.", host.Name())
+	}
+	return nil
+}
+
+func flattenVsanDisks(d *schema.ResourceData, cluster *object.ClusterComputeResource, client *govmomi.Client) error {
+	diskMap := []interface{}{}
+
+	hosts, err := clustercomputeresource.Hosts(cluster)
+	if err != nil {
+		return err
+	}
+	for _, host := range hosts {
```

Comment:
Random thought... I wonder what's the average size of this list in real world. I'm wondering if it would make sense to make some of these loops in parallel.